src/writer.js

```diff
@@ -7,32 +7,55 @@ const encode = require('./encode')
 const tracerVersion = require('../lib/version')
 
 class Writer {
-  constructor (url, size) {
+  constructor (url, size, processors) {
     this._queue = []
     this._url = url
     this._size = size
+    this._processors = processors
   }
 
   get length () {
     return this._queue.length
   }
 
+  // A trace is considered filtered out if it is undefined, or if it contains no spans
+  isTraceFilteredOut (trace) {
+    return trace === undefined || trace === []
+  }
+
+  applyProcessors (trace) {
+    return this._processors.reduce((currentTrace, processor) => {
```

Comment:
A valid trace is pretty much an array (of spans). A valid span should have the following structure minimally:

* `trace_id`: UInt64BE instance
* `span_id`: UInt64BE instance
* `name`: String
* `resource`: String
* `service`: String
* `start`: Int64BE instance
* `duration`: Int64BE instance

If `parent_id` is set, it should be a UInt64BE instance
If `error` is set, it should be a Number of `0` or `1`
If `meta` is set, it should be an object containing string:string key/value pairs
If `type` is set, it should be a string

I guess we could say that "format" becomes a pre-processor and there should now be an additional post-processor. In that case we could be a bit more flexible (i.e. allowing other types in certain cases like duration could be a Number and convert it back in the post-processor). This would also change the IDs to use buffers and start/duration to Numbers in the pre-processor, and only convert them to the appropriate type in the post-processor.